src/v/cloud_storage/inventory/inv_consumer.cc

```diff
@@ -0,0 +1,259 @@
+    auto h = _gate.hold();
+    report_parser parser{std::move(s), is_report_compressed};
+    co_await parser
+      .consume([this](auto&& paths) { return process_paths(std::move(paths)); })
+      .finally([&parser] { return parser.stop(); });
+    co_await flush(write_all_t::yes);
+}
+
+std::vector<ss::sstring> inventory_consumer::parse_row(ss::sstring row) {
+    std::vector<ss::sstring> pieces;
+    for (auto p : row | views::split(',') | views::transform(trim_string)) {
+        pieces.emplace_back(p);
+    }
+    return pieces;
+}
+
+ss::future<>
+inventory_consumer::process_paths(fragmented_vector<ss::sstring> paths) {
+    for (const auto& path : paths) {
+        // The row is expected to be in the form:
+        // "bucket","path"
+        auto pieces = parse_row(path);
+        // If a row is malformed we skip it and do not abort processing.
+        // This way the rest of the entries are processed, and if there
+        // was a path in this row it will be recorded as missing.
+        if (pieces.size() != 2) {
+            vlog(cst_log.warn, "unexpected row in inventory report: {}", path);
+        }
+        process_path(pieces[1]);
+    }
+
+    if (_total_size >= _max_hash_size_in_memory) {
+        co_await flush();
+    }
+}
+
+void inventory_consumer::process_path(ss::sstring path) {
+    if (auto maybe_ntp = ntp_from_path(path);
+        _ntps.contains(maybe_ntp.value())) {
+        const auto& ntp = maybe_ntp.value();
+        auto hash = xxhash_64(path.data(), path.size());
+
+        if (!_ntp_flush_states.contains(ntp)) {
+            _ntp_flush_states.insert({ntp, flush_state{}});
+        }
+
+        _ntp_flush_states[ntp].hashes.push_back(hash);
+        _total_size += sizeof(hash);
+    }
+}
+
+ss::future<>
+inventory_consumer::flush(inventory_consumer::write_all_t write_all_entries) {
+    auto h = _gate.hold();
+    vlog(
+      cst_log.trace,
+      "Flushing to disk, write_all_entries: {}",
+      write_all_entries);
+
+    _num_flushes += 1;
+    if (write_all_entries) {
+        for (auto& [ntp, flush_state] : _ntp_flush_states) {
+            if (!flush_state.hashes.empty()) {
+                co_await flush_ntp_hashes(
+                  ntp,
+                  std::move(flush_state.hashes),
+                  flush_state.next_file_name);
+            }
+        }
+        co_return;
+    }
+
+    // Set up a vector to sort entries. The ntp and next filename are not moved
+    // out of the map, only the hashes are moved out.
+    fragmented_vector<flush_entry> entries;
+    ranges::transform(
+      _ntp_flush_states, std::back_inserter(entries), [](auto& e) {
+          return flush_entry{
+            e.first, std::move(e.second.hashes), e.second.next_file_name};
+      });
+
+    ranges::sort(entries, ranges::greater(), [](const auto& e) {
+        return e.hashes.size();
+    });
+
+    const auto target_size = _max_hash_size_in_memory / 3;
```

Comment:
nit: why 3? Could you add a comment?

Reply:
sorry missed this before. I initially used half here but changed to 3 to make sure we did not have to dump data to disk too frequently, there isn't a strong reasoning behind this tbh. Maybe 1/2 would be fine.